bayesian/api/api_v2.py

```diff
@@ -172,6 +178,90 @@ def get(ecosystem, package, version):
+    """Implements stack analysis REST APIs.
+
+    Implements /api/v2/stack-analyses REST APIs for POST and GET calls.
+    """
+
+    method_decorators = [login_required]
+
+    @staticmethod
+    def get(external_request_id):
+        """Handle /api/v2/stack-analyses GET REST API."""
+        current_app.logger.debug("GET request_id: {}".format(external_request_id))
+
+        # 1. Set stack analyses object
+        sa = StackAnalyses()
+
+        # 2. Read request data for the request
+        db_result = sa.get_request_data(external_request_id)
+        if db_result is None:
+            error_message = "Invalid request ID '{}'.".format(external_request_id)
+            current_app.logger.error(error_message)
+            raise HTTPError(404, error=error_message)
+
+        # 3. Read stack results for the request
+        stack_result = sa.get_stack_result(external_request_id)
+
+        # 4. Read recommendation data for the request
+        recm_data = sa.get_recommendation_data(external_request_id)
+
+        # 5. If stack or reco is missing, either we are in progress or we have timedout.
+        if stack_result is None or recm_data is None:
+            # If the response is not ready and the timeout period is over, send error 408
+            if request_timed_out(db_result):
+                error_message = "Stack analysis request {} has timed out. Please retry " \
+                                "with a new analysis.".format(external_request_id)
+                current_app.logger.error(error_message)
+                raise HTTPError(408, error=error_message)
+            else:
+                error_message = "Analysis for request ID '{}' is in progress".format(
+                    external_request_id)
+                current_app.logger.error(error_message)
+                return {'error': error_message}, 202
+
+        # 6. Some rare case when workers have not updated the result
+        if stack_result == -1 and recm_data == -1:
+            error_message = "Worker result for request ID '{}' doesn't exist yet".format(
+                external_request_id)
+            current_app.logger.error(error_message)
+            raise HTTPError(404, error=error_message)
+
+        # 7. Assemble final GET response and return
+        sa_response_builder = StackAnalysesResponseBuilder(external_request_id,
+                                                           stack_result, recm_data)
+        return sa_response_builder.get_response()
+
+    @staticmethod
+    def post():
+        """Handle /api/v2/stack-analyses POST REST API."""
+        # 1. Read mandatory params
+        manifest = request.files.get("manifest") or None
+        file_path = request.form.get("file_path") or None
+        ecosystem = request.form.get("ecosystem") or None
+
+        current_app.logger.debug("Mandatory params :: manifest: {} file_path: {} "
+                                 "ecosystem: {}".format(manifest, file_path, ecosystem))
```

Comment:
Assigning default values can be part of the `StackAnalyses` request validation.

Reply:
You mean to create a default values for each parameter should be read from StackAnalyses class?

Reply:
If they are mandatory can we put an Assert and throw an exception if they are not present

Reply:
Yes we can do that, but I also validated there data type, so instead of doing too many things at API class, I have moved this responsibility to stack analyses class. I am Ok to move all this checks in API class as well, if you insist. Please comment if you need this movement.

Reply:
It will be good if you can add an Assert

Reply:
I've used `pydantic` to validate & assign request/response default values. It is quite easy to use and doesn't have a learning curve at all. `Pydantic` facilities declarative way to validate input/output data format.

https://github.com/fabric8-analytics/f8a-server-backbone/blob/ed78b2ce41067c29a963a8423260b8f710099b6e/src/v2/models.py#L224

Reply:
This is implemented in commit 'Used pydantic to validated post params'.